ext/js/language/translator.js

```diff
@@ -438,57 +443,90 @@ export class Translator {
+        const sourceCache = new Map(); // For reusing text processors' outputs
+
+        for (
+            let i = text.length;
+            i > 0;
+            i = this._getNextSubstringLength(options.searchResolution, i, text)
+        ) {
+            const rawSource = text.substring(0, i);
+
+            for (const preprocessorVariant of preprocessorVariants) {
+                let source = rawSource;
+
+                const textReplacements = /** @type {import('translation').FindTermsTextReplacement[] | null} */ (preprocessorVariant.get('textReplacements'));
+                if (textReplacements !== null) {
+                    source = this._applyTextReplacements(source, textReplacements);
+                }
 
-        for (const arrayVariant of this._generateArrayVariants(variantSpace)) {
-            const textReplacements = /** @type {import('translation').FindTermsTextReplacement[] | null} */ (arrayVariant.get('textReplacements'));
+                source = this._applyTextProcessors(textPreprocessors, preprocessorVariant, source, sourceCache);
 
-            let text2 = text;
-            const sourceMap = new TextSourceMap(text2);
+                if (used.has(source)) { continue; }
+                used.add(source);
+                for (const deinflection of this._multiLanguageTransformer.transform(language, source)) {
+                    const {trace, conditions} = deinflection;
+                    for (const postprocessorVariant of postprocessorVariants) {
+                        let {text: transformedText} = deinflection;
+                        transformedText = this._applyTextProcessors(textPostprocessors, postprocessorVariant, transformedText, sourceCache);
+
+                        /** @type {import('dictionary').InflectionRuleChainCandidate} */
+                        const inflectionRuleChainCandidate = {
+                            source: 'algorithm',
+                            inflectionRules: trace.map((frame) => frame.transform)
+                        };
+                        deinflections.push(this._createDeinflection(rawSource, source, transformedText, conditions, [inflectionRuleChainCandidate]));
+                    }
+                }
+            }
+        }
+        return deinflections;
+    }
 
-            if (textReplacements !== null) {
-                text2 = this._applyTextReplacements(text2, sourceMap, textReplacements);
+    /**
+     * @param {import('language').TextProcessorWithId<unknown>[]} textProcessors
+     * @param {Map<string, unknown>} processorVariant
+     * @param {string} text
+     * @param {Map<string, import('core').SafeAny>} textCache
+     * @returns {string}
+     */
+    _applyTextProcessors(textProcessors, processorVariant, text, textCache) {
+        for (const {id, textProcessor: {process}} of textProcessors) {
+            const setting = processorVariant.get(id);
+            let level1 = textCache.get(text);
```

Comment:
This textCache logic seems to be what sourceMap was doing originally. Is that correct?

Reply:
No, `sourceMap` had a functional purpose, it was necessary for the algorithm. The `textCache` is there just to improve performance by minimizing calls to textProcessor.process.